### ISSUE TITLE
list_aggs hangs sometimes hangs indefinitely

### ISSUE THREAD

**Issue by evanvolgas** [polygon-io/client-python#532]
**Describe the bug**
The `list_aggs` method in the Python Client will sometimes hang indefinitely without any response code or errors. It is not consistent about when it hangs and when it does not. There are times everything works correctly. 

**To Reproduce**
Run [this script](https://gist.github.com/evanvolgas/6a5cf5f9524a667cf0a7bae57ea0ed1e) using your Polygon API credentials (I assume in this script they are stored as an environment variable)

**Expected behavior**
This script runs successfully sometimes and assembles the dataframe of all stock prices. Often times, however, it instead proceeds through the dictionary of ticker/title pairs for about an hour before hanging indefinitely on a `polygon_client.list_aggs` call for a particular ticker. The ticker where the loop hangs is never the same.  When I look in the Polygon dashboard online, I don't see any errors in my API calls. This whole thing works and works and works until it stops working, at which point `polygon_client.list_aggs` doesn't throw an error, it just hangs.

**Comment by justinpolygon** [comment]
Hi @evanvolgas, I'm sorry to hear you're experiencing an inconsistency with the `list_aggs` method and hopefully we work together to get to the root of the issue. For a more in-depth look at what might be causing the hang, it might be beneficial to enable request debugging. By doing this, you can trace the underlying API calls and gather more insight. You can turn on debugging as outlined in our [official documentation](https://github.com/polygon-io/client-python#debugging-with-restclient). In short:

```python
client = RESTClient(trace=True)
```

Once enabled, this will output the request and response details, including a `request-id` which can be useful for us to trace the request on our side. In addition to this, it would be helpful if you could provide some context:

- How frequently do you encounter this issue? (Is it a matter of days, weeks?)
- Roughly, how many requests precede the hang?

If possible, please include some additional debugging logs immediately before and after the API calls to Polygon, so we can pinpoint exactly where the program gets stuck. I imagine you're doing something with the API results after you fetch them so I just wanted to make sure there wasn't a hang there too.

**Comment by evanvolgas** [comment]
Thank you for getting back to me so quickly :) 

I'll enable trace =True in just a moment and see if I can get more info. It typically takes an hour and a half for a successful run and an hour for a failed one, so I'll circle back once in an hour with some more info here. 

- **How frequently do you encounter this issue? (Is it a matter of days, weeks?)**. Approximately 75% of the time. 
- **How many requests am I making, in total?** 10,879 of them, exactly. (The total number of SEC tickers [here](https://www.sec.gov/files/company_tickers.json)
- **Roughly, how many requests precede the hang?**  Usually about 7000 of them, spread out like the image below. Note that this image is from a hung run that hung after exactly 9,815 requests. (the two errors in the log there are referring to aborted runs that occur after this particular run got hung). 
<img width="406" alt="Untitled" src="https://github.com/polygon-io/client-python/assets/3965137/d8dfa82d-2204-4be0-9c0f-b0ec4d65bdbc">

**Comment by evanvolgas** [comment]
```
Retrieving stocks data for DGX // QUEST DIAGNOSTICS INC
Request URL: https://api.polygon.io/v2/aggs/ticker/DGX/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 18:41:13 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '3776007dc01e0489e9c4da8e4da6efca', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for DGX // QUEST DIAGNOSTICS INC
Retrieving stocks data for NTRS // NORTHERN TRUST CORP
Request URL: https://api.polygon.io/v2/aggs/ticker/NTRS/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 18:41:16 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '702ebb33bd702af602631c5a9bb41494', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for NTRS // NORTHERN TRUST CORP
Retrieving stocks data for EPAM // EPAM Systems, Inc.
Request URL: https://api.polygon.io/v2/aggs/ticker/EPAM/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 18:41:17 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': 'b0ab6e9ec7f731234afa089ec401f841', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
---------------------------------------------------------------------------
_InactiveRpcError                         Traceback (most recent call last)
File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/grpc_helpers.py:75, in _wrap_unary_errors.<locals>.error_remapped_callable(*args, **kwargs)
     74 try:
---> 75     return callable_(*args, **kwargs)
     76 except grpc.RpcError as exc:

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/grpc/_channel.py:1161, in _UnaryUnaryMultiCallable.__call__(self, request, timeout, metadata, credentials, wait_for_ready, compression)
   1155 (
   1156     state,
   1157     call,
   1158 ) = self._blocking(
   1159     request, timeout, metadata, credentials, wait_for_ready, compression
   1160 )
-> 1161 return _end_unary_response_blocking(state, call, False, None)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/grpc/_channel.py:1004, in _end_unary_response_blocking(state, call, with_call, deadline)
   1003 else:
-> 1004     raise _InactiveRpcError(state)

_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "UNKNOWN:Deadline Exceeded {created_time:"2023-10-09T14:47:44.720763-04:00", grpc_status:4}"
>

The above exception was the direct cause of the following exception:

DeadlineExceeded                          Traceback (most recent call last)
File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/retry.py:204, in retry_target(target, predicate, sleep_generator, timeout, on_error, **kwargs)
    203 try:
--> 204     return target()
    206 # pylint: disable=broad-except
    207 # This function explicitly must deal with broad exceptions.

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/timeout.py:120, in TimeToDeadlineTimeout.__call__.<locals>.func_with_timeout(*args, **kwargs)
    118     kwargs["timeout"] = max(0, self._timeout - time_since_first_attempt)
--> 120 return func(*args, **kwargs)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/grpc_helpers.py:77, in _wrap_unary_errors.<locals>.error_remapped_callable(*args, **kwargs)
     76 except grpc.RpcError as exc:
---> 77     raise exceptions.from_grpc_error(exc) from exc

DeadlineExceeded: 504 Deadline Exceeded

The above exception was the direct cause of the following exception:

RetryError                                Traceback (most recent call last)
Cell In[3], line 30
     28 df = pd.DataFrame(daily)
     29 df['date'] = df['timestamp'].apply(lambda x: dt.datetime.fromtimestamp(x/1000))
---> 30 logger.log_text(f'Assembled dataframe of stock history for {ticker} // {title}', )
     31 print(f'Assembled dataframe of stock history for {ticker} // {title}')
     33 dfs.append(df)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py:199, in Logger.log_text(self, text, client, **kw)
    185 def log_text(self, text, *, client=None, **kw):
    186     """Log a text message
    187
    188     See
   (...)
    197             See :class:`~logging_v2.entries.LogEntry`.
    198     """
--> 199     self._do_log(client, TextEntry, text, **kw)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py:168, in Logger._do_log(self, client, _entry_class, payload, **kw)
    166     google.cloud.logging_v2._instrumentation_emitted = True
    167 # partial_success is true to avoid dropping instrumentation logs
--> 168 client.logging_api.write_entries(entries, partial_success=True)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/cloud/logging_v2/_gapic.py:163, in _LoggingAPI.write_entries(self, entries, logger_name, resource, labels, partial_success, dry_run)
    154 log_entry_pbs = [_log_entry_mapping_to_pb(entry) for entry in entries]
    156 request = WriteLogEntriesRequest(
    157     log_name=logger_name,
    158     resource=resource,
   (...)
    161     partial_success=partial_success,
    162 )
--> 163 self._gapic_api.write_log_entries(request=request)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/cloud/logging_v2/services/logging_service_v2/client.py:727, in LoggingServiceV2Client.write_log_entries(self, request, log_name, resource, labels, entries, retry, timeout, metadata)
    724 rpc = self._transport._wrapped_methods[self._transport.write_log_entries]
    726 # Send the request.
--> 727 response = rpc(
    728     request,
    729     retry=retry,
    730     timeout=timeout,
    731     metadata=metadata,
    732 )
    734 # Done; return the response.
    735 return response

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py:131, in _GapicCallable.__call__(self, timeout, retry, compression, *args, **kwargs)
    128 if self._compression is not None:
    129     kwargs["compression"] = compression
--> 131 return wrapped_func(*args, **kwargs)

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/retry.py:366, in Retry.__call__.<locals>.retry_wrapped_func(*args, **kwargs)
    362 target = functools.partial(func, *args, **kwargs)
    363 sleep_generator = exponential_sleep_generator(
    364     self._initial, self._maximum, multiplier=self._multiplier
    365 )
--> 366 return retry_target(
    367     target,
    368     self._predicate,
    369     sleep_generator,
    370     self._timeout,
    371     on_error=on_error,
    372 )

File ~/.pyenv/versions/3.10.4/envs/ds/lib/python3.10/site-packages/google/api_core/retry.py:220, in retry_target(target, predicate, sleep_generator, timeout, on_error, **kwargs)
    216     next_attempt_time = datetime_helpers.utcnow() + datetime.timedelta(
    217         seconds=sleep
    218     )
    219     if deadline < next_attempt_time:
--> 220         raise exceptions.RetryError(
    221             "Deadline of {:.1f}s exceeded while calling target function".format(
    222                 timeout
    223             ),
    224             last_exc,
    225         ) from last_exc
    227 _LOGGER.debug(
    228     "Retrying due to {}, sleeping {:.1f}s ...".format(last_exc, sleep)
    229 )
    230 time.sleep(sleep)

RetryError: Deadline of 60.0s exceeded while calling target function, last exception: 504 Deadline Exceeded
```

**Comment by justinpolygon** [comment]
Okay, thanks for letting me know. Yeah, I was wondering about that since the error message seemed to imply it already retried on the 504. Using a higher retry count like 10 might help. Let me know how it works out. Ideally, you'll never see these 504 type errors and we can dig into it on the support side if you have the request id's for the ones that fail.

**Comment by evanvolgas** [comment]
So it's not hanging anymore, but it is erroring out. Am I being rate limited? 

```
Retrieving dividend data for VIV // TELEFONICA BRASIL S.A.
Request URL: https://api.polygon.io/v3/reference/dividends?ticker=VIV
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:46 GMT', 'Content-Type': 'application/json', 'Content-Length': '544', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAyMi0wNC0yNyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '91756e4cf7614f8fabc3eeb207f77f74', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAyMi0wNC0yNyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:46 GMT', 'Content-Type': 'application/json', 'Content-Length': '579', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAyMC0xMi0yOSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '398885100eefaee9d79f6a5b9ee8d2c6', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAyMC0xMi0yOSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:46 GMT', 'Content-Type': 'application/json', 'Content-Length': '580', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxOS0wMy0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '8fc7295848f9b420db1551f8ce8d8a7e', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxOS0wMy0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:46 GMT', 'Content-Type': 'application/json', 'Content-Length': '580', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNy0wMy0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '92fb949d637cb6ef140f93f25a344932', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNy0wMy0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:47 GMT', 'Content-Type': 'application/json', 'Content-Length': '562', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNS0xMS0wMyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '872a0871b28b19f603bd6a7e74a41e9c', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNS0xMS0wMyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:47 GMT', 'Content-Type': 'application/json', 'Content-Length': '589', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNC0xMC0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '34ea906a360b16cd9fa4b0da63714db4', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxNC0xMC0wMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:10:47 GMT', 'Content-Type': 'application/json', 'Content-Length': '588', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxMy0wMS0yMiZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '035496c07b64388f40840a9700e31899', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAxMy0wMS0yMiZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:30:28 GMT', 'Content-Type': 'application/json', 'Content-Length': '630', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwOS0wMy0yNiZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '8688fb8678e0ac6051393fc63a0ae457', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwOS0wMy0yNiZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:30:28 GMT', 'Content-Type': 'application/json', 'Content-Length': '604', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwNi0xMS0xNyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '2a99e673614549409437077a78e0fb41', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwNi0xMS0xNyZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:30:28 GMT', 'Content-Type': 'application/json', 'Content-Length': '605', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Link': '<https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwMy0wNC0xMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg>; rel="next"', 'Vary': 'Accept-Encoding', 'X-Item-Count': '10', 'X-Request-Id': '00b51bba93937098784e6982075abd50', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Request URL: https://api.polygon.io/v3/reference/dividends?cursor=YXA9MjAwMy0wNC0xMSZhcz1WSVYmbGltaXQ9MTAmb3JkZXI9ZGVzYyZzb3J0PWV4X2RpdmlkZW5kX2RhdGUmdGlja2VyPVZJVg
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 20:30:28 GMT', 'Content-Type': 'application/json', 'Content-Length': '388', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Item-Count': '8', 'X-Request-Id': '6839c21216f22e2fdbf0a8b849ca31b7', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Retrived dividend data VIV for TELEFONICA BRASIL S.A.
Traceback (most recent call last):
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 75, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/grpc/_channel.py", line 1161, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/grpc/_channel.py", line 1004, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "recvmsg:Operation timed out"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"recvmsg:Operation timed out", grpc_status:14, created_time:"2023-10-09T16:43:56.796924-04:00"}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 204, in retry_target
    return target()
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 77, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.ServiceUnavailable: 503 recvmsg:Operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/evan/Documents/new/gh/arapbi/new-life/polygon_scraper_dividends.py", line 38, in <module>
    logger.log_text(f'Retrieving dividend data for {ticker} // {title}', )
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py", line 199, in log_text
    self._do_log(client, TextEntry, text, **kw)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py", line 168, in _do_log
    client.logging_api.write_entries(entries, partial_success=True)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/_gapic.py", line 163, in write_entries
    self._gapic_api.write_log_entries(request=request)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/services/logging_service_v2/client.py", line 727, in write_log_entries
    response = rpc(
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 366, in retry_wrapped_func
    return retry_target(
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 220, in retry_target
    raise exceptions.RetryError(
google.api_core.exceptions.RetryError: Deadline of 60.0s exceeded while calling target function, last exception: 503 recvmsg:Operation timed out
```


This crapped out on request 1225 on my end. A concurrent dividend scraped crapped out on request 653 on my end. 

Is what I am trying to do unreasonable? It doesn't feel like I'm hitting your API that hard, am I?

**Comment by evanvolgas** [comment]
Here's an example of a hung job. It hung on request 3,969 out of 10,879.

```
Retrieving stocks data for ACB // AURORA CANNABIS INC
Request URL: https://api.polygon.io/v2/aggs/ticker/ACB/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 21:06:04 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '1cf160278b84a6d16ebb975c40b951fd', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for ACB // AURORA CANNABIS INC
Retrieving stocks data for FUND // SPROTT FOCUS TRUST INC.
Request URL: https://api.polygon.io/v2/aggs/ticker/FUND/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 21:06:05 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '52cfb32c23c3dd54279ff7c9d1128f26', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for FUND // SPROTT FOCUS TRUST INC.
Retrieving stocks data for NYXH // Nyxoah SA
Request URL: https://api.polygon.io/v2/aggs/ticker/NYXH/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
```

It's like the request is made, none is ever received, and instead of crapping out, it just hangs there indefinitely. 

I'm confused why it would be crapping out at all, let alone hang indefinitely.

**Comment by evanvolgas** [comment]
I've tried resubmitting the script a few times and I keep getting the behavior above: everything is working fine until eventually a request is made and no response is received. The loop never completes because the job gets stuck after making several `polygon_client.list_aggs` calls. Any advice or thoughts you might have would be appreciated. I'm not sure what to do next on my end to debug further.

**Comment by justinpolygon** [comment]
> google.api_core.exceptions.RetryError: Deadline of 60.0s exceeded while calling target function, last exception: 503 recvmsg:Operation timed out

Hi @evanvolgas,  read through all the comments and these looks like google API 503 errors now and nothing to do with Polygon. Have you looked into your GCP logs? It's totally reasonable to fetch dividends from Polygon. After you make a request to polygon and you print the len(results) or something to make sure you're actually getting a result. Just since you seem to have a few things going on I'd like to make sure it's actually a polygon issue and not after you received the results and then the next step in the process is hanging.

These all seem to be google libs related now:

Traceback (most recent call last):
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 204, in retry_target
    return target()
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/grpc_helpers.py", line 77, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.ServiceUnavailable: 503 recvmsg:Operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/evan/Documents/new/gh/arapbi/new-life/polygon_scraper_dividends.py", line 38, in <module>
    logger.log_text(f'Retrieving dividend data for {ticker} // {title}', )
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py", line 199, in log_text
    self._do_log(client, TextEntry, text, **kw)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/logger.py", line 168, in _do_log
    client.logging_api.write_entries(entries, partial_success=True)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/_gapic.py", line 163, in write_entries
    self._gapic_api.write_log_entries(request=request)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/cloud/logging_v2/services/logging_service_v2/client.py", line 727, in write_log_entries
    response = rpc(
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 366, in retry_wrapped_func
    return retry_target(
  File "/Users/evan/.pyenv/versions/ds/lib/python3.10/site-packages/google/api_core/retry.py", line 220, in retry_target
    raise exceptions.RetryError(
google.api_core.exceptions.RetryError: Deadline of 60.0s exceeded while calling target function, last exception: 503 recvmsg:Operation timed out

**Comment by evanvolgas** [comment]
I deleted the earlier log because I noticed the same thing you did but only after I pasted it here. 

GCP timeouts are a separate issue. Agreed. Note that the [script](https://gist.github.com/evanvolgas/6a5cf5f9524a667cf0a7bae57ea0ed1e) I'm using has been updated to remove the GCP logging calls. What I'm seeing now, running the exact script I linked to, is the behavior I described [here](https://github.com/polygon-io/client-python/issues/532#issuecomment-1754015447) eg,

```
Retrieving stocks data for ACB // AURORA CANNABIS INC
Request URL: https://api.polygon.io/v2/aggs/ticker/ACB/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 21:06:04 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '1cf160278b84a6d16ebb975c40b951fd', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for ACB // AURORA CANNABIS INC
Retrieving stocks data for FUND // SPROTT FOCUS TRUST INC.
Request URL: https://api.polygon.io/v2/aggs/ticker/FUND/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Mon, 09 Oct 2023 21:06:05 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '52cfb32c23c3dd54279ff7c9d1128f26', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for FUND // SPROTT FOCUS TRUST INC.
Retrieving stocks data for NYXH // Nyxoah SA
Request URL: https://api.polygon.io/v2/aggs/ticker/NYXH/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
```

The last request to https://api.polygon.io/v2/aggs/ticker/NYXH/range/1/day/2018-01-01/2023-12-31?limit=50000 never responds. It just hangs indefinitely. I provided a few example API calls that succeeded from the logfile as well. 3,968 of them succeeded. This scripts hung on request 3,969 out of 10,879 in this particular invocation. I could run it again and it would hang somewhere else.

**Comment by evanvolgas** [comment]
Another curious matter, but perhaps a different bug, is that it appears you only have data for 9,793 of the 10,879 tickers that the [SEC identifies](https://www.sec.gov/files/company_tickers.json). The job just ran successfully. I'll try running it again to see if it keeps succeeding. It's very sporadic when it works and when (and where) it fails though.

**Comment by justinpolygon** [comment]
Hi @evanvolgas, I was just looking through the SEC json file that you're using to seed tickers and for some there isn't a 1-1 mapping to polygon. We have the Tickers API where you can query all the ticker symbols https://polygon.io/docs/stocks/get_v3_reference_tickers. So, this might be a cause of some issues too in they you're likely asking for data that doesn't exist.

For example from the SEC file:

```
    {
        "cik_str": 1067983,
        "ticker": "BRK-B",
        "title": "BERKSHIRE HATHAWAY INC"
    },
```

Then on the polygon side these would be:

```
{
    "results": [
        {
            "ticker": "BRK.A",
            "name": "Berkshire Hathaway Inc.",
            "market": "stocks",
            "locale": "us",
            "primary_exchange": "XNYS",
            "type": "CS",
            "active": true,
            "currency_name": "usd",
            "cik": "0001067983",
            "composite_figi": "BBG000DWCFL4",
            "share_class_figi": "BBG001S902J2",
            "last_updated_utc": "2023-10-09T00:00:00Z"
        },
        {
            "ticker": "BRK.B",
            "name": "BERKSHIRE HATHAWAY Class B",
            "market": "stocks",
            "locale": "us",
            "primary_exchange": "XNYS",
            "type": "CS",
            "active": true,
            "currency_name": "usd",
            "cik": "0001067983",
            "composite_figi": "BBG000DWG505",
            "share_class_figi": "BBG001S90346",
            "last_updated_utc": "2023-10-09T00:00:00Z"
        }
    ],
    "status": "OK",
    "request_id": "c010cbb661516cec616e0c2564b19501",
    "count": 2
}
```

You might have better luck using the https://polygon.io/docs/stocks/get_v3_reference_tickers API.

**Comment by evanvolgas** [comment]
Got it. I'll refactor to use the `Tickers API ` Thank you.

I'm still seeing unusual hangs with no log messages to indicate what happened except that a request was made and never answered. It hangs on me about 75% of the time. 

It looks like I can refactor my approach to hit the `get_grouped_daily_aggs` endpoint over multiple days. I'll also write a job that hits `get_grouped_daily_aggs` to see if those calls hang indefinitely on me as well.

**Comment by evanvolgas** [comment]
This does not freeze when I hit `get_grouped_daily_aggs` instead, which is fine. Still, it's a little concerning that `list_aggs` hangs. I'll keep trying to track it down on my end, but it feels like what's happening is that occasionally a requests is getting swallowed on Polygon's side, which in turns causes the client to hang while waiting for a response.

**Comment by justinpolygon** [comment]
Can you maybe try wrapping the ticker for loop with something like this:

```
for ticker, title in tickers.items():
    print(f'Retrieving stocks data for {ticker} // {title}')
    try:
        # ... rest of the loop for fetching and processing data ...
    except Exception as e:
        print(f"Error processing ticker {ticker}: {e}")
```

This is a tough one since I cannot reproduce it. Can you reproduce this locally on your dev machine?

**Comment by evanvolgas** [comment]
I've added that and am trying to reproduce the error.

**Comment by evanvolgas** [comment]
Same thing as before.  The last three API calls are

```
Retrieving stocks data for AMRC // Ameresco, Inc.
Request URL: https://api.polygon.io/v2/aggs/ticker/AMRC/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Tue, 10 Oct 2023 18:41:19 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '9eba28e8000fdaba3e7400154b7a4b6e', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for AMRC // Ameresco, Inc.
Retrieving stocks data for SKYW // SKYWEST INC
Request URL: https://api.polygon.io/v2/aggs/ticker/SKYW/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
Response Headers: {'Server': 'nginx/1.19.2', 'Date': 'Tue, 10 Oct 2023 18:41:19 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Vary': 'Accept-Encoding', 'X-Request-Id': '823d356d079888e877663ba6705598f2', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains'}
Assembled dataframe of stock history for SKYW // SKYWEST INC
Retrieving stocks data for WMK // WEIS MARKETS INC
Request URL: https://api.polygon.io/v2/aggs/ticker/WMK/range/1/day/2018-01-01/2023-12-31?limit=50000
Request Headers: {'Authorization': 'Bearer REDACTED', 'Accept-Encoding': 'gzip', 'User-Agent': 'Polygon.io PythonClient/1.12.8'}
```

No response is provided and no exception is thrown.

**Comment by justinpolygon** [comment]
Very strange. You're seeing the same error on your local machine? Can you send me a script so that I can reproduce it?

**Comment by evanvolgas** [comment]
Certainly.

```
import datetime as dt

import pandas as pd
import pandas_gbq
import requests

from google.cloud import logging, secretmanager, storage
from polygon import RESTClient

# Constants
BQ_TABLE_NAME = 'raw.all_tickers_history' # dataset.table
BUCKET_NAME = "arapbi-polygon"
COMPANY_TICKERS_URL = 'https://www.sec.gov/files/company_tickers.json'
CSV_FILE_NAME = "all_tickers_history.csv"
GCP_FOLDER_NAME = "polygon/tickers/"
GCP_FILE_NAME = GCP_FOLDER_NAME + CSV_FILE_NAME
LOG_NAME = "polygon_all_tickers_history"
PROJECT_ID = "new-life-400922"
SECRET_ID = "polygon"

# Set up client connections
# Rewrite this with https://cloud.google.com/logging/docs/reference/libraries#client-libraries-usage-python
logging_client = logging.Client()
logger = logging_client.logger(LOG_NAME)
secrets_client = secretmanager.SecretManagerServiceClient()
polygon_secret = secrets_client.access_secret_version(request={"name": f"projects/{PROJECT_ID}/secrets/{SECRET_ID}/versions/latest"})
polygon_client = RESTClient(polygon_secret.payload.data.decode("UTF-8"), retries=10, trace=True)
storage_client = storage.Client()

# Assemlble a dictionary of stock tickers from the SEC's listing of them
r = requests.get(COMPANY_TICKERS_URL)
tickers_df = pd.DataFrame(dict(r.json())).T
tickers = tickers_df.set_index('ticker')['title'].to_dict()

# Scrape Polygon's website for stocks history for every ticker, make a dataframe out of the result,
# and append that dataframe to a list of all dataframes for all stocks. It will be concatenated to one dataframe below
dfs = []
for ticker, title in tickers.items():
    # logger.log_text(f'Retrieving stocks data for {ticker} // {title}')
    print(f'Retrieving stocks data for {ticker} // {title}')
    try:
        aggs = []
        for a in polygon_client.list_aggs(
            ticker,
            1,
            "day",
            "2018-01-01",
            "2023-12-31",
            limit=50000,
        ):
            aggs.append(a)

        if aggs:
            daily  = [{ 'ticker': ticker,
                        'title': title,
                        'timestamp': y.timestamp,
                        'open': y.open,
                        'close': y.close,
                        'volume_weighted_average_price': y.vwap,
                        'volume': y.volume,
                        'transactions': y.transactions } for y in aggs]
            df = pd.DataFrame(daily)
            df['date'] = df['timestamp'].apply(lambda x: dt.datetime.fromtimestamp(x/1000))
            # logger.log_text(f'Assembled dataframe of stock history for {ticker} // {title}', )
            print(f'Assembled dataframe of stock history for {ticker} // {title}')

            dfs.append(df)
        else:
            # logger.log_text(f'No stocks data for {ticker} on behalf of {title}')
            print(f'No stocks data for {ticker} on behalf of {title}')
    except Exception as e:
        print(f"Error processing ticker {ticker}: {e}")

# Create the master dataframe of all stocks and their price history
# logger.log_text(f'Making stocks DataFrame')
print(f'Making stocks DataFrame')
all_tickers_history = pd.concat(dfs)
all_tickers_history = all_tickers_history.drop('timestamp', axis = 1)
all_tickers_history['ticker'] = all_tickers_history['ticker'].astype(str)
all_tickers_history['title'] = all_tickers_history['title'].astype(str)
all_tickers_history['open'] = all_tickers_history['open'].fillna(0).astype(float)
all_tickers_history['close'] = all_tickers_history['close'].fillna(0).astype(float)
all_tickers_history['volume_weighted_average_price'] = all_tickers_history['volume_weighted_average_price'].fillna(0).astype(float)
all_tickers_history['volume'] = all_tickers_history['volume'].fillna(0).astype(int)
all_tickers_history['transactions'] = all_tickers_history['transactions'].fillna(0).astype(int)
all_tickers_history['date'] = pd.to_datetime(all_tickers_history['date'])

# Write a CSV
# logger.log_text(f'Writing stocks CSV')
print(f'Writing CSV')
all_tickers_history.to_csv(CSV_FILE_NAME)

# Upload the CSV to Google Storage
# logger.log_text(f'Uploading stocks file')
print(f'Uploading file')
bucket = storage_client.bucket(BUCKET_NAME)
blob = bucket.blob(GCP_FILE_NAME)
blob.upload_from_filename(CSV_FILE_NAME, timeout=3600)

# upload the data to Bigquery
# logger.log_text(f'Uploading to BQ')
print(f'Uploading to BQ')
pandas_gbq.to_gbq(all_tickers_history,
                  BQ_TABLE_NAME,
                  project_id=PROJECT_ID,
                  if_exists='replace',
                  table_schema=[{'name': 'ticker','type': 'STRING'},
                                {'name': 'title','type': 'STRING'},
                                {'name': 'open','type': 'FLOAT'},
                                {'name': 'close','type': 'FLOAT'},
                                {'name': 'volume_weighted_average_price','type': 'FLOAT'},
                                {'name': 'volume','type': 'INT64'},
                                {'name': 'transactions','type': 'INT64'},
                                {'name': 'date','type': 'DATE'}
                                ],)
```

**Comment by justinpolygon** [comment]
Hey, I stripped out the google stuff and was able to run the following without issue. Yeah, you really want to be using the Tickers API to get the correct ticker symbols. Have a second run running right now and I'll see if I can get it to throw an error.

Here's the code:

```python
import datetime as dt

import pandas as pd
import requests

from polygon import RESTClient

COMPANY_TICKERS_URL = 'https://www.sec.gov/files/company_tickers.json'
CSV_FILE_NAME = "all_tickers_history.csv"
LOG_NAME = "polygon_all_tickers_history"

polygon_client = RESTClient(retries=10, trace=True)

# Assemlble a dictionary of stock tickers from the SEC's listing of them
r = requests.get(COMPANY_TICKERS_URL)
tickers_df = pd.DataFrame(dict(r.json())).T
tickers = tickers_df.set_index('ticker')['title'].to_dict()

# Scrape Polygon's website for stocks history for every ticker, make a dataframe out of the result,
# and append that dataframe to a list of all dataframes for all stocks. It will be concatenated to one dataframe below
dfs = []
for ticker, title in tickers.items():
    # logger.log_text(f'Retrieving stocks data for {ticker} // {title}')
    print(f'Retrieving stocks data for {ticker} // {title}')
    try:
        aggs = []
        for a in polygon_client.list_aggs(
            ticker,
            1,
            "day",
            "2018-01-01",
            "2023-12-31",
            limit=50000,
        ):
            aggs.append(a)

        if aggs:
            daily  = [{ 'ticker': ticker,
                        'title': title,
                        'timestamp': y.timestamp,
                        'open': y.open,
                        'close': y.close,
                        'volume_weighted_average_price': y.vwap,
                        'volume': y.volume,
                        'transactions': y.transactions } for y in aggs]
            df = pd.DataFrame(daily)
            df['date'] = df['timestamp'].apply(lambda x: dt.datetime.fromtimestamp(x/1000))
            # logger.log_text(f'Assembled dataframe of stock history for {ticker} // {title}', )
            print(f'Assembled dataframe of stock history for {ticker} // {title}')

            dfs.append(df)
        else:
            # logger.log_text(f'No stocks data for {ticker} on behalf of {title}')
            print(f'No stocks data for {ticker} on behalf of {title}')
    except Exception as e:
        print(f"Error processing ticker {ticker}: {e}")

# Create the master dataframe of all stocks and their price history
print(f'Making stocks DataFrame')
all_tickers_history = pd.concat(dfs)
all_tickers_history = all_tickers_history.drop('timestamp', axis = 1)
all_tickers_history['ticker'] = all_tickers_history['ticker'].astype(str)
all_tickers_history['title'] = all_tickers_history['title'].astype(str)
all_tickers_history['open'] = all_tickers_history['open'].fillna(0).astype(float)
all_tickers_history['close'] = all_tickers_history['close'].fillna(0).astype(float)
all_tickers_history['volume_weighted_average_price'] = all_tickers_history['volume_weighted_average_price'].fillna(0).astype(float)
all_tickers_history['volume'] = all_tickers_history['volume'].fillna(0).astype(int)
all_tickers_history['transactions'] = all_tickers_history['transactions'].fillna(0).astype(int)
all_tickers_history['date'] = pd.to_datetime(all_tickers_history['date'])

# Write a CSV
print(f'Writing CSV')
all_tickers_history.to_csv(CSV_FILE_NAME)
```

Here's any updated version that runs much quicker because we can use ThreadPoolExecutor to run about 50 requests in parallel so it completes much much faster.

```python
import datetime as dt
import pandas as pd
import requests
from polygon import RESTClient
from concurrent.futures import ThreadPoolExecutor, as_completed
from threading import Lock

COMPANY_TICKERS_URL = 'https://www.sec.gov/files/company_tickers.json'
CSV_FILE_NAME = "all_tickers_history.csv"
LOG_NAME = "polygon_all_tickers_history"
WORKERS = 50

polygon_client = RESTClient(retries=10, trace=True)

# Get tickers from SEC's listing
r = requests.get(COMPANY_TICKERS_URL)
tickers_df = pd.DataFrame(dict(r.json())).T
tickers = tickers_df.set_index('ticker')['title'].to_dict()

# Print the count of tickers
print(f"Total tickers found: {len(tickers)}")

dfs = []
dfs_lock = Lock()  # Lock for ensuring thread safety when appending to dfs

def fetch_stock_history(ticker_title):
    ticker, title = ticker_title
    print(f'Retrieving stocks data for {ticker} // {title}')
    try:
        aggs = []
        for a in polygon_client.list_aggs(ticker, 1, "day", "2018-01-01", "2023-12-31", limit=50000):
            aggs.append(a)

        if aggs:
            daily = [{'ticker': ticker,
                      'title': title,
                      'timestamp': y.timestamp,
                      'open': y.open,
                      'close': y.close,
                      'volume_weighted_average_price': y.vwap,
                      'volume': y.volume,
                      'transactions': y.transactions} for y in aggs]
            df = pd.DataFrame(daily)
            df['date'] = df['timestamp'].apply(lambda x: dt.datetime.fromtimestamp(x/1000))
            print(f'Assembled dataframe of stock history for {ticker} // {title}')
            
            with dfs_lock:
                dfs.append(df)
        else:
            print(f'No stocks data for {ticker} on behalf of {title}')
    except Exception as e:
        print(f"Error processing ticker {ticker}: {e}")

# Using ThreadPoolExecutor to fetch stock histories concurrently
with ThreadPoolExecutor(max_workers=WORKERS) as executor:
    executor.map(fetch_stock_history, tickers.items())

print(f'Making stocks DataFrame')
all_tickers_history = pd.concat(dfs)
all_tickers_history = all_tickers_history.drop('timestamp', axis=1)
all_tickers_history['ticker'] = all_tickers_history['ticker'].astype(str)
all_tickers_history['title'] = all_tickers_history['title'].astype(str)
all_tickers_history['open'] = all_tickers_history['open'].fillna(0).astype(float)
all_tickers_history['close'] = all_tickers_history['close'].fillna(0).astype(float)
all_tickers_history['volume_weighted_average_price'] = all_tickers_history['volume_weighted_average_price'].fillna(0).astype(float)
all_tickers_history['volume'] = all_tickers_history['volume'].fillna(0).astype(int)
all_tickers_history['transactions'] = all_tickers_history['transactions'].fillna(0).astype(int)
all_tickers_history['date'] = pd.to_datetime(all_tickers_history['date'])

print(f'Writing CSV')
all_tickers_history.to_csv(CSV_FILE_NAME)
```

This write a 625MB csv file with 7,875,598 lines.

**Comment by evanvolgas** [comment]
it does hold it in memory, but I have plenty to spare on my local and I'll be sizing my VM's appropriately. 

Please give me a day or two to circle back here while I do some more testing myself. I'm not abandoning this, just need a bit of time to poke at it some more.

**Comment by evanvolgas** [comment]
Good news and (maybe) bad news. 

Good news first. I have not hit this behavior again. I have no idea what the heck was going on there, but I can't recreate it anymore. I'm going to close this and I'll reopen it if I start seeing the behavior again. I'll be running daily so I'll notice it if there's a problem. 

(Maybe) bad news: Because the reference/tickers endpoint includes [duplicate ticker ids](https://github.com/polygon-io/issues/issues/272#issuecomment-1760544659), you'd need to dedup them before you could do something [like this](https://github.com/polygon-io/client-python/issues/532#issuecomment-1754669675). Mentioning to you as an FYI cause it was unexpected for me (it may not be for you?).
